File: fix_final_syntax.py
#!/usr/bin/env python3
"""
Final syntax fixer for specific issues
"""
import os
import re
import glob

def fix_final_syntax(content):
    """Fix final syntax errors"""
    # Fix 'use client' directive placement
    content = re.sub(r"'use client',\s*", "'use client';\n", content)
    
    # Fix missing commas in object properties
    lines = content.split('\n')
    fixed_lines = []
    
    for i, line in enumerate(lines):
        # Fix missing commas in object properties
        if re.match(r'^\s*[a-zA-Z_][a-zA-Z0-9_]*:\s*[\'"][^\'"]*[\'"]$', line.strip()):
            # Check if next line is not a closing brace and doesn't start with comma
            if i + 1 < len(lines):
                next_line = lines[i + 1].strip()
                if next_line and not next_line.startswith('}') and not next_line.startswith(','):
                    # Add comma if missing
                    if not line.strip().endswith(','):
                        line = line.rstrip() + ','
        
        # Fix array elements missing commas
        if re.match(r'^\s*[\'"][^\'"]*[\'"]$', line.strip()):
            # Check if this is in an array context
            if i + 1 < len(lines):
                next_line = lines[i + 1].strip()
                if next_line and not next_line.startswith(']') and not next_line.startswith(','):
                    # Add comma if missing
                    if not line.strip().endswith(','):
                        line = line.rstrip() + ','
        
        # Fix broken className attributes
        line = re.sub(r'className="([^"]*?)([a-z])([A-Z])', r'className="\1\2 \3', line)
        line = re.sub(r'className="([^"]*?)([a-z])([a-z])([A-Z])', r'className="\1\2\3 \4', line)
        
        # Fix broken JSX structure
        line = re.sub(r'<div className="([^"]+)" />([^<]+)</div>', r'<div className="\1">\2</div>', line)
        line = re.sub(r'<h1 className="([^"]+)" />([^<]+)</h1>', r'<h1 className="\1">\2</h1>', line)
        line = re.sub(r'<p className="([^"]+)" />([^<]+)</p>', r'<p className="\1">\2</p>', line)
        
        # Fix missing semicolons in import statements
        if line.strip().startswith('import ') and not line.strip().endswith(';'):
            line = line.rstrip() + ';'
        
        # Fix trailing commas in object declarations
        line = re.sub(r'export const metadata: Metadata = {,', 'export const metadata: Metadata = {', line)
        line = re.sub(r'openGraph: {,', 'openGraph: {', line)
        
        fixed_lines.append(line)
    
    return '\n'.join(fixed_lines)

def fix_file(file_path):
    """Fix final syntax errors in a single file"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        original_content = content
        content = fix_final_syntax(content)
        
        if content != original_content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(content)
            return True
        
        return False
        
    except Exception as e:
        print(f"Error fixing {file_path}: {e}")
        return False

def main():
    """Main function to fix all final syntax errors"""
    patterns = [
        './app/**/*.tsx',
        './app/**/*.ts', 
        './app/**/*.js',
        './app/**/*.jsx',
        './src/**/*.tsx',
        './src/**/*.ts',
        './src/**/*.js', 
        './src/**/*.jsx',
        './components/**/*.tsx',
        './components/**/*.ts',
        './components/**/*.js',
        './components/**/*.jsx'
    ]
    
    files_to_fix = []
    for pattern in patterns:
        files_to_fix.extend(glob.glob(pattern, recursive=True))
    
    fixed_count = 0
    
    for file_path in files_to_fix:
        if os.path.isfile(file_path):
            if fix_file(file_path):
                print(f"Fixed: {file_path}")
                fixed_count += 1
    
    print(f"\nFixed {fixed_count} files")

if __name__ == "__main__":
    main()

File: fix_final_syntax.js
#!/usr/bin/env node;
import fs from 'fs';'import path from 'path''// Get all TypeScript/TSX files in the app directory
function getAllTsFiles(dir) {;
let files = [];
const items = fs.readdirSync(dir)
  for (const item of items) {;
const fullPath = path.join(dir, item);
const stat = fs.statSync(fullPath)
    if (stat.isDirectory()) {
  
      files = files.concat(getAllTsFiles(fullPath))
    } else if (item.endsWith('.tsx') || item.endsWith('.ts')) {''      files.push(fullPath)
    }
  }

  return files
}

function fixFinalSyntaxErrors(filePath) {
  
  try {;
let content = fs.readFileSync(filePath, 'utf8');';'const originalContent = content
    // Fix specific syntax issues

    // 1. Fix missing commas in object arrays
    content = content.replace(/benefits:\s*\[[^\]]*\]\s*}\s*$/gm, (match) => {
  
      if (!match.includes(',')) {''        return match.replace(/}\s*$/, '},')''      }
      return match
    })
    // 2. Fix missing commas after object properties
    content = content.replace(/description:\s*'[^']*',\s*benefits:\s*\[[^\]]*\]\s*}\s*$/gm, (match) => {''      return match.replace(/}\s*$/, '},')''    })
    // 3. Fix missing semicolons after const declarations
    content = content.replace(/(\s+const\s+\w+\s*=\s*[^;]+)\s*(\n\s*const|\n\s*return|\n\s*function|\n\s*export)/gm, (match, declaration, nextPart) => {
  
      if (!declaration.trim().endsWith(') && !declaration.trim().endsWith('}') && !declaration.trim().endsWith(')')) {''        return declaration + ' + nextPart''      }
      return match
    })
    // 4. Fix missing closing brackets for arrays
    content = content.replace(/(\s+const\s+\w+\s*=\s*\[[\s\S]*?)(\s*)(\n\s*const|\n\s*return|\n\s*function|\n\s*export)/gm, (match, arrayPart, spaces, nextPart) => {
  
      if (!arrayPart.trim().endsWith(']') && !arrayPart.trim().endsWith('],')) {''        return arrayPart + ']' + spaces + nextPart''      }
      return match
    })
    // 5. Fix malformed JSX return statements
    content = content.replace(/return\s*\(\s*<>\s*$/gm, 'return (\n    <>')''    // 6. Fix missing closing tags
    content = content.replace(/<main[^>]*>[\s\S]*?<\/section>\s*$/gm, (match) => {
  
      if (!match.includes('</main>')) {''        return match + '\n      </main>\n    </>\n  );\n}'''      }
      return match
    })
    // 7. Fix missing export statements;
    if (content.includes('const ') && !content.includes('export default') && !content.includes('export { ')) {;';'const componentName = content.match(/const\s+(\w+):\s*React\.FC/g)
      if (componentName) {
  
        content = content.replace(/}\s*$/, `}\n\nexport default ${componentName[1]}`)
      }
    }

    // 8. Fix extra semicolons in object properties
    content = content.replace(/,\s*;/g, ',')''    // 9. Fix missing commas in object arrays
    content = content.replace(/}\s*\n\s*]\s*$/gm, '}\n  ]')''    // 10. Fix missing commas after array declarations
    content = content.replace(/(\s+const\s+\w+\s*=\s*\[[\s\S]*?\])\s*(\n\s*const|\n\s*return|\n\s*function|\n\s*export)/gm, (match, arrayPart, nextPart) => {
  
      if (!arrayPart.trim().endsWith(')) {''        return arrayPart + ' + nextPart''      }
      return match
    })
    // 11. Fix missing commas in object properties
    content = content.replace(/description:\s*'[^']*',\s*benefits:\s*\[[^\]]*\]\s*}\s*$/gm, (match) => {''      return match.replace(/}\s*$/, '},')''    })
    // 12. Fix missing semicolons after array declarations
    content = content.replace(/(\s+const\s+\w+\s*=\s*\[[\s\S]*?\])\s*(\n\s*const|\n\s*return|\n\s*function|\n\s*export)/gm, (match, arrayPart, nextPart) => {
  
      if (!arrayPart.trim().endsWith(')) {''        return arrayPart + ' + nextPart''      }
      return match
    })
    // Clean up extra whitespace
    content = content.replace(/\n\n\n+/g, '\n\n')''    content = content.trim() + '\n'''    if (content !== originalContent) {
  
      fs.writeFileSync(filePath, content, 'utf8')''      console.log(`✓ Fixed final syntax errors in: ${filePath}`)
      return true
    }

    return false
  } catch (error) {
  
    console.error(`✗ Error processing ${filePath}:`, error.message)
    return false
  }
}

// Get all TypeScript files;
const tsFiles = getAllTsFiles('./app')''console.log(`Found ${tsFiles.length} TypeScript files to check for final syntax errors`);
let fixedCount = 0;
let errorCount = 0
tsFiles.forEach(filePath => {
  
  if (fs.existsSync(filePath)) {
  
    if (fixFinalSyntaxErrors(filePath)) {
  
      fixedCount++
    }
  } else {
  
    console.log(`⚠ File not found: ${filePath}`)
  }
})
console.log(`\nSummary:`)
console.log(`- Files processed: ${tsFiles.length}`)
console.log(`- Files fixed: ${fixedCount}`)
console.log(`- Errors: ${errorCount}`)